quinn-proto/src/connection.rs

```diff
@@ -957,22 +982,201 @@ impl Connection {
+        remote: SocketAddrV6,
+        partial_decode: PartialDecode,
+    ) -> Option<BytesMut> {
+        match partial_decode.finish() {
+            Ok((packet, rest)) => {
+                self.handle_packet(ctx, now, remote, packet);
+                rest
+            }
+            Err(e) => {
+                trace!(ctx.log, "unable to complete packet decoding"; "reason" => %e);
+                None
+            }
+        }
+    }
+
+    fn handle_packet(
         &mut self,
         ctx: &mut Context,
         now: u64,
         remote: SocketAddrV6,
         mut packet: Packet,
+    ) {
+        if let Some(token) = self.params.stateless_reset_token {
+            if packet.payload.len() >= 16 && packet.payload[packet.payload.len() - 16..] == token {
+                if !self.state.as_ref().unwrap().is_drained() {
+                    debug!(ctx.log, "got stateless reset"; "connection" => %self.loc_cid);
+                    ctx.io.push_back(Io::TimerStop {
+                        connection: self.handle,
+                        timer: Timer::LossDetection,
+                    });
+                    ctx.io.push_back(Io::TimerStop {
+                        connection: self.handle,
+                        timer: Timer::Close,
+                    });
+                    ctx.io.push_back(Io::TimerStop {
+                        connection: self.handle,
+                        timer: Timer::Idle,
+                    });
+                    ctx.events.push_back((
+                        self.handle,
+                        Event::ConnectionLost {
+                            reason: ConnectionError::Reset,
+                        },
+                    ));
+                    self.state = Some(State::Drained);
+                }
+                return;
+            }
+        }
+
+        trace!(ctx.log, "connection got packet"; "connection" => %self.loc_cid, "len" => packet.payload.len());
+        let (prev_state, was_handshake) = match self.state.take().unwrap() {
+            State::Handshake(mut state) => {
+                if !state.rem_cid_set {
+                    match packet.header {
+                        Header::Long {
+                            src_cid: rem_cid, ..
+                        } => {
+                            trace!(ctx.log, "got remote connection id"; "connection" => %self.loc_cid, "rem_cid" => %rem_cid);
+                            self.rem_cid = rem_cid;
+                            state.rem_cid_set = true;
+                        }
+                        _ => {}
+                    }
+                }
+                (State::Handshake(state), true)
+            }
+            state => (state, false),
+        };
+        let was_closed = prev_state.is_closed();
+
+        let result = match self.decrypt_packet(was_handshake, &mut packet) {
+            Ok(number) => {
+                match prev_state {
+                    State::Handshake(_) | State::Established => {
```

Comment:
This is equivalent to branching on `was_closed`.

Reply:
Ah good, that's a bit more concise.